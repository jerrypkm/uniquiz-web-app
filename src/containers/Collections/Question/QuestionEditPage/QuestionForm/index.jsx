import React, { lazy } from "react";

import { useForm } from "react-hook-form";
import { withRouter } from "react-router-dom";
import { compose } from "recompose";
import { useCallback } from "react";
import Button from "@material-ui/core/Button";
import MoPageHeaderEdit from "components/library/MoPageHeaderEdit";
import MoSpinner from "components/library/MoSpinner";
import MoHintEdit from "components/library/MoHintEdit";
import MoPageContentEdit from "components/library/MoPageContentEdit";
import styles from "./styles";
import withStyles from "@material-ui/core/styles/withStyles";
import Box from "@material-ui/core/Box";

const CodeEditor = lazy(() => import("components/shared/CodeEditor"));

const QuestionForm = ({
  classes,
  children,
  isLoading,
  label,
  question,
  subtitle,
  setQuestion,
  title,
  onSubmit,
}) => {
  const { handleSubmit, register } = useForm();

  const handleChange = useCallback(
    (code) => {
      const userAnswer = { ...Object.values(code) };
      if (userAnswer[0] === "{}" || userAnswer[0] === typeof Object) {
        return;
      }
      setQuestion({ ...question, ...code });
    },
    [setQuestion, question]
  );

  if (isLoading || !question) {
    return <MoSpinner isLoading={isLoading} color="primary" />;
  }

  return (
    <form onSubmit={handleSubmit((e) => onSubmit({ ...question, ...e }))}>
      <section className={classes.section}>
        <MoPageHeaderEdit
          text={title}
          register={register}
          name="title"
          placeholder={"Html Paragraph"}
        />
        <MoPageContentEdit
          text={label}
          register={register}
          name="label"
          placeholder={"Create a paragraph open tag in html"}
        />
        <MoHintEdit
          text={subtitle}
          register={register}
          name="subtitle"
          placeholder={"<p>"}
        />
      </section>
      <section className={classes.section}>
        <CodeEditor
          codeAnswer={"Write Question Here"}
          codeLanguage={question?.language}
          codeQuestion={question?.question}
          isConsole={true}
          isEditMode={true}
          noInline={false}
          isPlayground={question?.isPlayground}
          handleOnChange={(code) => handleChange({ question: code.userAnswer })}
          sm={6}
          md={6}
          title="Question Code"
        />
      </section>
      {question && (
        <section className={classes.section}>
          <CodeEditor
            codeAnswer={"Answer"}
            codeLanguage={question?.language}
            codeQuestion={question?.answer}
            isConsole={true}
            isEditMode={true}
            noInline={false}
            isPlayground={question?.isPlayground}
            handleOnChange={(code) => handleChange({ answer: code.userAnswer })}
            sm={6}
            md={6}
            title="Answer Code"
          />
        </section>
      )}

      {children && <section className={classes.section}>{children}</section>}
      {question && (
        <Box position="static" width="100%" zIndex="10">
          <Button
            onKeyPress={(e) => {
              e.key === "Enter" && e.preventDefault();
            }}
            type="submit"
            color="primary"
            variant="contained"
          >
            Save
          </Button>
        </Box>
      )}
    </form>
  );
};

export default compose(withRouter, withStyles(styles))(QuestionForm);
