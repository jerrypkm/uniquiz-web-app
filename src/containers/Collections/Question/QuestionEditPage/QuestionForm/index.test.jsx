import React from "react";

import ShallowRenderer from "react-test-renderer/shallow";
import QuestionForm from ".";

const renderer = new ShallowRenderer();

describe("QuestionForm Component", () => {
  test("should match the snapshot", () => {
    renderer.render(<QuestionForm />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
