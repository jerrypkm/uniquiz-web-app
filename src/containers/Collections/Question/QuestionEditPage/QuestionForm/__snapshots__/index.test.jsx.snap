// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionForm Component should match the snapshot 1`] = `
<Context.Consumer>
  [Function]
</Context.Consumer>
`;
